narwhal/executor/src/subscriber.rs

```diff
@@ -243,151 +275,311 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
     /// requests to remote worker if not found locally
-    #[instrument(level = "debug", skip_all, fields(digest = % digest, worker_id = % worker_id))]
-    async fn fetch_payload(
+    async fn fetch_batches_from_worker(
         &self,
-        digest: BatchDigest,
-        worker_id: WorkerId,
+        batch_digests: HashMap<WorkerId, HashSet<BatchDigest>>,
         workers: Vec<NetworkPublicKey>,
-    ) -> Batch {
-        if let Some(payload) = self.try_fetch_locally(digest, worker_id).await {
-            let batch_fetch_duration = payload.metadata.created_at.elapsed().as_secs_f64();
-            self.metrics
-                .batch_execution_latency
-                .observe(batch_fetch_duration);
+    ) -> HashMap<BatchDigest, Batch> {
+        let mut fetched_batches = HashMap::new();
+
+        for (worker_id, digests) in batch_digests {
             debug!(
-                "Batch {:?} took {} seconds to be fetched for execution since creation",
-                payload.digest(),
-                batch_fetch_duration
+                "Attempting to fetch {} digests from {} worker_{worker_id}'s",
+                digests.len(),
+                workers.len()
             );
-            return payload;
-        }
-        let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
-        let mut stagger = Duration::from_secs(0);
-        let mut futures = vec![];
-        for worker in workers {
-            let future = self.fetch_from_worker(stagger, worker, digest);
-            futures.push(future.boxed());
-            // TODO: Make this a parameter, and also record workers / authorities that are down
-            //       to request from them batches later.
-            stagger += Duration::from_millis(200);
+            // TODO: Can further parallelize this by worker_id if necessary.
+            // Only have one worker for now so will leave this for a future
+            // optimization.
+            let mut remaining_digests = digests.clone();
+
+            let local_batches = self
+                .try_fetch_locally(remaining_digests.clone().into_iter().collect(), worker_id)
+                .await;
+            for (local_batch_digest, local_batch) in local_batches {
+                if remaining_digests.remove(&local_batch_digest) {
+                    let batch_fetch_duration =
+                        local_batch.metadata.created_at.elapsed().as_secs_f64();
+                    self.metrics
+                        .batch_execution_latency
+                        .observe(batch_fetch_duration);
+                    debug!(
+                        "Batch {local_batch_digest:?} took {batch_fetch_duration} seconds to be fetched for execution since creation",
+                    );
+                    fetched_batches.insert(local_batch_digest, local_batch);
+                }
+            }
+
+            if remaining_digests.is_empty() {
+                return fetched_batches;
+            }
+
```

Comment:
Maybe have a loop to chunk remaining_digests per 1000 items?

Reply:
If I do chunk it here, would the idea be to issue those chunks in parallel or is this related to your idea of limiting outside of the handler? Wonder if it would saturate the network too much with so many potentially large responses being returned at the same time?

Reply:
I think chunking by # at the requestor side is cleaner. Limiting total size has to be done at the handler though. Since this is a trusted interface between primary and worker, I think we can divide the responsibilities?